sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/QueryBuilder.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.mongodb;
+
+import com.mongodb.BasicDBObject;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoCursor;
+import com.mongodb.client.model.Aggregates;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.Projections;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.bson.BsonDocument;
+import org.bson.Document;
+import org.bson.codecs.BsonValueCodecProvider;
+import org.bson.codecs.IterableCodecProvider;
+import org.bson.codecs.ValueCodecProvider;
+import org.bson.codecs.configuration.CodecRegistries;
+import org.bson.conversions.Bson;
+import org.bson.types.ObjectId;
+
+/**
+ * Builds an AggregateIterable pipeline using multiple options.
+ *
+ * @author Ahmed Elhossaini
+ */
+public class QueryBuilder {
+  private Integer limit;
+  private List<String> projection;
+  private List<BsonDocument> pipeline;
+  MongoCollection<Document> collection;
+  private String filter;
+  private String documentIdStr;
+  private ObjectId documentId;
+
+  public static QueryBuilder create(MongoCollection<Document> collection) {
+    QueryBuilder builder = new QueryBuilder();
+    builder.collection = collection;
+
+    return builder;
+  }
+
+  public QueryBuilder withDocumentIdStr(String documentId) {
+    this.documentIdStr = documentId;
+
+    return this;
+  }
+
+  public QueryBuilder withDocumentId(ObjectId documentId) {
+    this.documentId = documentId;
+
+    return this;
+  }
+
+  public QueryBuilder withLimit(Integer limit) {
+    this.limit = limit;
+
+    return this;
+  }
+
+  public QueryBuilder withProjection(List<String> projection) {
+    this.projection = projection;
+
+    return this;
+  }
+
+  public QueryBuilder withPipeline(List<BsonDocument> pipeline) {
+    this.pipeline = pipeline;
+
+    return this;
+  }
+
+  public QueryBuilder withFilter(String filter) {
+    this.filter = filter;
+
+    return this;
+  }
+
+  public MongoCursor<Document> cursor() {
+    List<BsonDocument> mongoDbPipeline = null;
+    // A custom pipeline is supplied
+    if (this.pipeline != null && !this.pipeline.isEmpty()) {
```

Comment:
I would suggest that you enforce that the user either sets `pipeline` or the other bits, but not both. Silently overwriting would be confusing. If you can merge them (like adding filters and stuff to the end of `this.pipeline`) that is cool too.

Reply:
Makes a lot of sense, I'll modify the code accordingly.